cmd/stunstamp/stunstamp.go

```diff
@@ -218,10 +239,17 @@ func nodeMetaFromDERPMap(dm *tailcfg.DERPMap, nodeMetaByAddr map[netip.Addr]node
 	return stale, nil
 }
 
-func getStableConns(stableConns map[netip.Addr][2]io.ReadWriteCloser, addr netip.Addr) ([2]io.ReadWriteCloser, error) {
-	conns, ok := stableConns[addr]
+func getStableConns(stableConns map[netip.Addr]map[int][2]io.ReadWriteCloser, addr netip.Addr, dstPort int) ([2]io.ReadWriteCloser, error) {
+	conns := [2]io.ReadWriteCloser{}
+	byDstPort, ok := stableConns[addr]
 	if ok {
-		return conns, nil
+		conns, ok = byDstPort[dstPort]
+		if ok {
+			return conns, nil
+		}
+	}
+	if byDstPort == nil {
+		byDstPort = make(map[int][2]io.ReadWriteCloser)
```

Comment:
If I read it right, you could do the update `stableConns[addr] = byDstPort` here so that you only need to write back to the map the first time. Not that it will matter for performance, but my eyes tripped on the missing map update till I found it below 🙂

Reply:
Oh actually, perhaps instead: Move this down there, since there's no need to construct the map at all till we've survived the error checks below.

Reply:
updated in https://github.com/tailscale/tailscale/pull/12356/commits/dcfa8bfd2a96a4cefc638518fab88df49e5be625